### DISCUSSION TITLE
ShowWindow to only allow a single instance of a window

### DISCUSSION THREAD

**Discussion by alexhelms** [canton7/Stylet#271] [Q&A] [answered]
I'm looking for a suggestion on what you think the best way to do this is, and perhaps an API extension to the WindowManager's ShowWindow function.

I have a window that I would like to effectively be a singleton. The user should be able to open it and still interact with other windows, so it should not be a "dialog" where it is modal. If the window is already open, it should bring it to the front instead of opening another.

I tried in the window's view model to check Application.Current.Windows to see if it already exists, and if so, call RequestClose() but that doesn't seem to work and was throwing an exception.

What would be the Stylet way of doing this? Would this be a good thing for Stylet to handle with a potential new `ShowSingletonWindow` function?

**Comment by canton7** [comment, marked as answer]
I'd probably start with something like this (untested)

```cs
public class SingletonWindowManager
{
    private readonly IWindowManager windowManager;
    private readonly SingletonWindowViewModel viewModel;

    public SingletonWindowViewManager(IWindowManager windowManager, SingletonWindowViewModel viewModel) =>
        (this.windowManager, this.viewModel) = (windowManager, viewModel);

    public void EnsureVisible()
    {
        if (viewModel.View != null)
        {
            // It's shown, just bring it to the front
            ((Window)viewModel.View).Activate();
        }
        else
        {
            windowManager.ShowWindow(viewModel);
        }
    }
}
```

then register it as a singleton:

```cs
protected override void ConfigureIoC(StyletIoCBuilder builder)
{
    builder.Bind<SingletonWindowManager>().ToSelf().InSingletonScope();
}
```

**Reply by alexhelms** [reply]
That is a great approach and worked great for my needs, thank you!

**Reply by canton7** [reply]
Glad to hear!
